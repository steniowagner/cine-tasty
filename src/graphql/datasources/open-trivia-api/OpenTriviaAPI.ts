import { RESTDataSource } from 'apollo-datasource-rest';

import { QuestionCategory, QuizInput, Question } from '@lib/types';
import shuffleArray from '@utils/shuffle-array/shuffleArray';
import { QuestionResponse } from '@open-trivia-api-types';

import {
  drawnTypeQuestionMixed,
  makeQueryString,
  makeURLParams,
  parseResult,
} from './helpers';

import CONSTANTS from './utils/constants';

type GetRequestResponse = {
  results: QuestionResponse[];
  response_code: number;
};

class OpenTriviaAPI extends RESTDataSource {
  constructor() {
    super();
    this.baseURL = CONSTANTS.BASE_URL;
  }

  async getQuestions(input: QuizInput): Promise<Question[]> {
    if (input.category.toLowerCase() === QuestionCategory.Mixed.toLowerCase()) {
      return this.getQuestionsMixedTypes({
        numberOfQuestions: input.numberOfQuestions,
        difficulty: input.difficulty,
        type: input.type,
      });
    }

    return this.getQuestionsSingleType(input);
  }

  async getQuestionsMixedTypes(input: Omit<QuizInput, 'category'>): Promise<Question[]> {
    if (input.numberOfQuestions === 1) {
      return this.getQuestionsSingleType({
        ...input,
        category: drawnTypeQuestionMixed(),
      });
    }

    const movieQuestionsAmount = Math.ceil(input.numberOfQuestions / 2);
    const tvQuestionsAmount = Math.floor(input.numberOfQuestions / 2);

    const [tvQuestions, movieQuestions] = await Promise.all<Question[], Question[]>([
      this.getQuestionsSingleType({
        ...input,
        category: QuestionCategory.Tv,
        numberOfQuestions: tvQuestionsAmount,
      }),
      this.getQuestionsSingleType({
        ...input,
        category: QuestionCategory.Movie,
        numberOfQuestions: movieQuestionsAmount,
      }),
    ]);

    return shuffleArray<Question>([...tvQuestions, ...movieQuestions]);
  }

  getRequestQueryString(input: QuizInput): string {
    const urlParams = makeURLParams(input);

    const queryString = makeQueryString(urlParams);

    return queryString;
  }

  async getQuestionsSingleType(input: QuizInput): Promise<Question[]> {
    const queryString = this.getRequestQueryString(input);

    const { results, response_code: responseCode } = await this.get<GetRequestResponse>(
      CONSTANTS.ENDPOINT,
      queryString,
    );

    if (responseCode === CONSTANTS.NO_RESPONSE_CODE) {
      return [];
    }

    return parseResult(results);
  }
}

export default OpenTriviaAPI;
