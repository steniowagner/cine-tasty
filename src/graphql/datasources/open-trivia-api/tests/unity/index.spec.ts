const mockRestDataSourceGet = jest.fn();

import {
  QuestionDifficulty,
  QuestionType,
  QuestionCategory,
} from '../../../../../lib/types';
import {
  question,
  movieQuestion,
  tvQuestion,
} from '../../../../../__tests__/mocks/quiz.stub';
import CONSTANTS from '../../utils/constants';
import OpenTriviaAPI from '../..';

jest.mock('apollo-datasource-rest', () => {
  class MockRESTDataSource {
    baseUrl = '';
    get = mockRestDataSourceGet;
  }

  return {
    RESTDataSource: MockRESTDataSource,
    HTTPCache: class HTTPCache {},
  };
});

const movieQuestionsInput = {
  difficulty: QuestionDifficulty.Hard,
  type: QuestionType.Multiple,
  category: QuestionCategory.Movie,
  number_questions: 10,
};

const tvQuestionsInput = {
  difficulty: QuestionDifficulty.Hard,
  type: QuestionType.Multiple,
  category: QuestionCategory.Tv,
  number_questions: 10,
};

const mixedQuestionsInput = {
  difficulty: QuestionDifficulty.Hard,
  type: QuestionType.Multiple,
  category: QuestionCategory.Mixed,
  number_questions: 2,
};

const triviaAPI = new OpenTriviaAPI();

describe('Unity: OpenTriviaAPI', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('getQuestions()', () => {
    describe('Getting questions about Movies', () => {
      it('should return an array with questions about movie from OpenTriviaAPI', async () => {
        mockRestDataSourceGet.mockReturnValue({ response_code: 0, results: [question] });

        const result = await triviaAPI.getQuestions(movieQuestionsInput);

        expect(mockRestDataSourceGet.mock.calls.length).toBe(1);

        expect(result).toMatchSnapshot();

        expect(mockRestDataSourceGet).toHaveBeenCalledWith(
          CONSTANTS.ENDPOINT,
          `category=${CONSTANTS.MOVIE_CATEGORY_CODE}&amount=${
            movieQuestionsInput.number_questions
          }&difficulty=${movieQuestionsInput.difficulty.toLowerCase()}&type=${movieQuestionsInput.type.toLowerCase()}`,
        );
      });

      it('should return an empty array when the OpenTriviaAPI returns the response_code = 1', async () => {
        mockRestDataSourceGet.mockReturnValue({
          response_code: CONSTANTS.NO_RESPONSE_CODE,
        });

        const result = await triviaAPI.getQuestions(movieQuestionsInput);

        expect(mockRestDataSourceGet.mock.calls.length).toBe(1);

        expect(result).toEqual([]);

        expect(mockRestDataSourceGet).toHaveBeenCalledWith(
          CONSTANTS.ENDPOINT,
          `category=${CONSTANTS.MOVIE_CATEGORY_CODE}&amount=${
            movieQuestionsInput.number_questions
          }&difficulty=${movieQuestionsInput.difficulty.toLowerCase()}&type=${movieQuestionsInput.type.toLowerCase()}`,
        );
      });
    });

    describe('Getting questions about Tv', () => {
      it('should return an array questions about tv from OpenTriviaAPI', async () => {
        mockRestDataSourceGet.mockReturnValue({ response_code: 0, results: [question] });

        const result = await triviaAPI.getQuestions(tvQuestionsInput);

        expect(mockRestDataSourceGet.mock.calls.length).toBe(1);

        expect(result).toMatchSnapshot();

        expect(mockRestDataSourceGet).toHaveBeenCalledWith(
          CONSTANTS.ENDPOINT,
          `category=${CONSTANTS.TV_CATEGORY_CODE}&amount=${
            tvQuestionsInput.number_questions
          }&difficulty=${tvQuestionsInput.difficulty.toLowerCase()}&type=${tvQuestionsInput.type.toLowerCase()}`,
        );
      });

      it('should return an empty array when the OpenTriviaAPI returns the response_code = 1', async () => {
        mockRestDataSourceGet.mockReturnValue({
          response_code: CONSTANTS.NO_RESPONSE_CODE,
        });

        const result = await triviaAPI.getQuestions(tvQuestionsInput);

        expect(mockRestDataSourceGet.mock.calls.length).toBe(1);

        expect(result).toEqual([]);

        expect(mockRestDataSourceGet).toHaveBeenCalledWith(
          CONSTANTS.ENDPOINT,
          `category=${CONSTANTS.TV_CATEGORY_CODE}&amount=${
            tvQuestionsInput.number_questions
          }&difficulty=${tvQuestionsInput.difficulty.toLowerCase()}&type=${tvQuestionsInput.type.toLowerCase()}`,
        );
      });
    });

    describe('Getting questions about Movies and TV', () => {
      it('should return an array with questions about movies and tv from OpenTriviaAPI', async () => {
        mockRestDataSourceGet
          .mockReturnValueOnce({ response_code: 0, results: [tvQuestion] })
          .mockReturnValueOnce({ response_code: 0, results: [movieQuestion] });

        const result = await triviaAPI.getQuestions(mixedQuestionsInput);

        expect(
          result.findIndex(item => item.category === tvQuestion.category),
        ).toBeGreaterThanOrEqual(0);

        expect(
          result.findIndex(item => item.category === movieQuestion.category),
        ).toBeGreaterThanOrEqual(0);

        expect(mockRestDataSourceGet.mock.calls.length).toBe(2);

        expect(mockRestDataSourceGet).toHaveBeenCalledWith(
          CONSTANTS.ENDPOINT,
          `category=${
            CONSTANTS.TV_CATEGORY_CODE
          }&amount=1&difficulty=${mixedQuestionsInput.difficulty.toLowerCase()}&type=${movieQuestionsInput.type.toLowerCase()}`,
        );

        expect(mockRestDataSourceGet).toHaveBeenCalledWith(
          CONSTANTS.ENDPOINT,
          `category=${CONSTANTS.MOVIE_CATEGORY_CODE}&amount=${Math.ceil(
            [movieQuestion].length / 2,
          )}&difficulty=${mixedQuestionsInput.difficulty.toLowerCase()}&type=${tvQuestionsInput.type.toLowerCase()}`,
        );
      });

      it('should return only tv questions when the movie request response return the response_code = 1', async () => {
        mockRestDataSourceGet
          .mockReturnValueOnce({ response_code: 0, results: [tvQuestion] })
          .mockReturnValueOnce({
            response_code: CONSTANTS.NO_RESPONSE_CODE,
          });

        const result = await triviaAPI.getQuestions(mixedQuestionsInput);

        expect(mockRestDataSourceGet.mock.calls.length).toBe(2);

        expect(result).toEqual([tvQuestion]);

        expect(mockRestDataSourceGet).toHaveBeenCalledWith(
          CONSTANTS.ENDPOINT,
          `category=${
            CONSTANTS.TV_CATEGORY_CODE
          }&amount=1&difficulty=${mixedQuestionsInput.difficulty.toLowerCase()}&type=${movieQuestionsInput.type.toLowerCase()}`,
        );

        expect(mockRestDataSourceGet).toHaveBeenCalledWith(
          CONSTANTS.ENDPOINT,
          `category=${CONSTANTS.MOVIE_CATEGORY_CODE}&amount=${Math.ceil(
            [movieQuestion].length / 2,
          )}&difficulty=${mixedQuestionsInput.difficulty.toLowerCase()}&type=${tvQuestionsInput.type.toLowerCase()}`,
        );
      });

      it('should return only movie questions when the tv request response return the response_code = 1', async () => {
        mockRestDataSourceGet
          .mockReturnValueOnce({
            response_code: CONSTANTS.NO_RESPONSE_CODE,
          })
          .mockReturnValueOnce({ response_code: 0, results: [movieQuestion] });

        const result = await triviaAPI.getQuestions(mixedQuestionsInput);

        expect(mockRestDataSourceGet.mock.calls.length).toBe(2);

        expect(result).toEqual([movieQuestion]);

        expect(mockRestDataSourceGet).toHaveBeenCalledWith(
          CONSTANTS.ENDPOINT,
          `category=${
            CONSTANTS.TV_CATEGORY_CODE
          }&amount=1&difficulty=${mixedQuestionsInput.difficulty.toLowerCase()}&type=${movieQuestionsInput.type.toLowerCase()}`,
        );

        expect(mockRestDataSourceGet).toHaveBeenCalledWith(
          CONSTANTS.ENDPOINT,
          `category=${CONSTANTS.MOVIE_CATEGORY_CODE}&amount=${Math.ceil(
            [movieQuestion].length / 2,
          )}&difficulty=${mixedQuestionsInput.difficulty.toLowerCase()}&type=${tvQuestionsInput.type.toLowerCase()}`,
        );
      });

      it('should return an empty array when both requests return response_code = 1', async () => {
        mockRestDataSourceGet
          .mockReturnValueOnce({
            response_code: CONSTANTS.NO_RESPONSE_CODE,
          })
          .mockReturnValueOnce({
            response_code: CONSTANTS.NO_RESPONSE_CODE,
          });

        const result = await triviaAPI.getQuestions(mixedQuestionsInput);

        expect(mockRestDataSourceGet.mock.calls.length).toBe(2);

        expect(result).toEqual([]);

        expect(mockRestDataSourceGet).toHaveBeenCalledWith(
          CONSTANTS.ENDPOINT,
          `category=${
            CONSTANTS.TV_CATEGORY_CODE
          }&amount=1&difficulty=${mixedQuestionsInput.difficulty.toLowerCase()}&type=${movieQuestionsInput.type.toLowerCase()}`,
        );

        expect(mockRestDataSourceGet).toHaveBeenCalledWith(
          CONSTANTS.ENDPOINT,
          `category=${CONSTANTS.MOVIE_CATEGORY_CODE}&amount=${Math.ceil(
            [movieQuestion].length / 2,
          )}&difficulty=${mixedQuestionsInput.difficulty.toLowerCase()}&type=${tvQuestionsInput.type.toLowerCase()}`,
        );
      });

      it('should return an array of question correctly when user want a single question fo mixed types', async () => {
        mockRestDataSourceGet
          .mockReturnValueOnce({ response_code: 0, results: [tvQuestion] })
          .mockReturnValueOnce({ response_code: 0, results: [movieQuestion] });

        const tvRequestURL = `category=${
          CONSTANTS.TV_CATEGORY_CODE
        }&amount=1&difficulty=${mixedQuestionsInput.difficulty.toLowerCase()}&type=${movieQuestionsInput.type.toLowerCase()}`;

        const movieRequestURL = `category=${
          CONSTANTS.MOVIE_CATEGORY_CODE
        }&amount=${Math.ceil(
          [movieQuestion].length / 2,
        )}&difficulty=${mixedQuestionsInput.difficulty.toLowerCase()}&type=${tvQuestionsInput.type.toLowerCase()}`;

        const result = await triviaAPI.getQuestions({
          ...mixedQuestionsInput,
          number_questions: 1,
        });

        expect(mockRestDataSourceGet.mock.calls.length).toBe(1);

        expect(result.length).toEqual(1);

        const isTVDrawn = JSON.stringify(tvQuestion) === JSON.stringify(result[0]);
        const isMovieDrawn = JSON.stringify(movieQuestion) === JSON.stringify(result[0]);

        const isTVURLCalled = mockRestDataSourceGet.mock.calls[0][1] === tvRequestURL;
        const isMovieURLCalled =
          mockRestDataSourceGet.mock.calls[0][1] === movieRequestURL;

        expect(mockRestDataSourceGet.mock.calls[0][0]).toEqual(CONSTANTS.ENDPOINT);

        expect(isTVURLCalled || isMovieURLCalled).toBe(true);

        expect(isTVDrawn || isMovieDrawn).toBe(true);
      });
    });
  });

  describe('getCategoryCode()', () => {
    it('should return the category code correctly for Movies', () => {
      const code = triviaAPI.getCategoryCode(QuestionCategory.Movie);

      expect(code).toEqual(CONSTANTS.MOVIE_CATEGORY_CODE);
    });

    it('should return the category code correctly for Tv', () => {
      const code = triviaAPI.getCategoryCode(QuestionCategory.Tv);

      expect(code).toEqual(CONSTANTS.TV_CATEGORY_CODE);
    });
  });

  describe('parseQueryParams()', () => {
    it('should return a query string correctly', () => {
      const queryParamsParsed = triviaAPI.parseQueryParams({
        category: 14,
        amount: 10,
        difficulty: 'difficulty',
        type: 'type',
      });

      expect(queryParamsParsed).toEqual(
        'category=14&amount=10&difficulty=difficulty&type=type',
      );
    });

    it('should return a query string correctly when the amount is greater than 50', () => {
      const queryParamsParsed = triviaAPI.parseQueryParams({
        category: 14,
        amount: 45,
        difficulty: 'difficulty',
        type: 'type',
      });

      expect(queryParamsParsed).toEqual(
        'category=14&amount=45&difficulty=difficulty&type=type',
      );
    });

    it('should return a query string correctly when the amount is less than 1', () => {
      const queryParamsParsed = triviaAPI.parseQueryParams({
        category: 14,
        amount: 1,
        difficulty: 'difficulty',
        type: 'type',
      });

      expect(queryParamsParsed).toEqual(
        'category=14&amount=1&difficulty=difficulty&type=type',
      );
    });
  });

  describe('getQuestionAmount()', () => {
    it('should return the question amount correctly when the amount is beteween [max, min] interval', () => {
      expect(triviaAPI.getQuestionAmount(10)).toEqual(10);
    });

    it('should return the max question amount when the input is greater than the max amount supported', () => {
      expect(triviaAPI.getQuestionAmount(CONSTANTS.MAX_QUESTIONS_REQUEST + 1)).toEqual(
        CONSTANTS.MAX_QUESTIONS_REQUEST,
      );
    });

    it('should return the min question amount when the input is less than the min amount supported', () => {
      expect(triviaAPI.getQuestionAmount(CONSTANTS.MIN_QUESTIONS_REQUEST - 1)).toEqual(
        CONSTANTS.MIN_QUESTIONS_REQUEST,
      );
    });
  });

  describe('getURLParams()', () => {
    it('should return the query params string correctly when all possible params are set', () => {
      const input = {
        number_questions: 10,
        difficulty: QuestionDifficulty.Hard,
        category: QuestionCategory.Movie,
        type: QuestionType.Multiple,
      };

      const urlParams = triviaAPI.getURLParams(input);

      expect(urlParams).toBe(
        `category=${
          CONSTANTS.MOVIE_CATEGORY_CODE
        }&amount=10&difficulty=${input.difficulty.toLowerCase()}&type=${input.type.toLowerCase()}`,
      );
    });

    it('should return the query params string correctly when some params have "any" as value', () => {
      let input = {
        number_questions: 10,
        difficulty: QuestionDifficulty.Easy,
        category: QuestionCategory.Movie,
        type: QuestionType.Mixed,
      };

      expect(triviaAPI.getURLParams(input)).toBe(
        `category=${
          CONSTANTS.MOVIE_CATEGORY_CODE
        }&amount=10&difficulty=${input.difficulty.toLowerCase()}`,
      );

      input = {
        number_questions: 10,
        difficulty: QuestionDifficulty.Mixed,
        category: QuestionCategory.Movie,
        type: QuestionType.Mixed,
      };

      expect(triviaAPI.getURLParams(input)).toBe(
        `category=${CONSTANTS.MOVIE_CATEGORY_CODE}&amount=10`,
      );
    });
  });
});
