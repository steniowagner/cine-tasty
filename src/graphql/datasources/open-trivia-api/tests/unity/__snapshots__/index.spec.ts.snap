// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unity: OpenTriviaAPI getQuestions() Getting questions about Movies should return an array with questions about movie from OpenTriviaAPI 1`] = `
Array [
  Object {
    "category": "category",
    "correct_answer": "correct_answer",
    "difficulty": "difficulty",
    "incorrect_answers": Array [
      "incorrect_answers",
    ],
    "question": "question",
    "type": "type",
  },
]
`;

exports[`Unity: OpenTriviaAPI getQuestions() Getting questions about Tv should return an array questions about tv from OpenTriviaAPI 1`] = `
Array [
  Object {
    "category": "category",
    "correct_answer": "correct_answer",
    "difficulty": "difficulty",
    "incorrect_answers": Array [
      "incorrect_answers",
    ],
    "question": "question",
    "type": "type",
  },
]
`;
